File: widget/lib/analytics/xyz.ts
/**
 * XYZ анализ — классификация по стабильности спроса (коэффициент вариации).
 *
 * Принимает массив периодных продаж по каждому SKU.
 * Класс X — CV ≤ 10% (стабильный спрос, можно работать "точно в срок").
 * Класс Y — CV 10–25% (сезонный/умеренно колеблющийся).
 * Класс Z — CV > 25% (нерегулярный, страховой запас обязателен).
 */

import { mean, stddev } from '../utils';

export type XyzInput = {
  id: string;
  name: string;
  periods: number[]; // продажи по периодам (дни/недели/месяцы)
};

export type XyzClass = 'X' | 'Y' | 'Z';

export type XyzRow = {
  id: string;
  name: string;
  mean: number;
  stddev: number;
  cv: number; // коэффициент вариации (0..∞)
  class: XyzClass;
  totalSales: number;
  periodsCount: number;
};

export type XyzOptions = {
  thresholds?: { x: number; y: number }; // по умолчанию x=0.10, y=0.25
};

export function buildXyzReport(inputs: XyzInput[], opts: XyzOptions = {}): XyzRow[] {
  const xT = opts.thresholds?.x ?? 0.1;
  const yT = opts.thresholds?.y ?? 0.25;

  return inputs.map((r) => {
    const xs = r.periods.filter(Number.isFinite);
    const m = mean(xs);
    const sd = stddev(xs);
    const cv = m > 0 ? sd / m : Infinity;
    const cls: XyzClass = cv <= xT ? 'X' : cv <= yT ? 'Y' : 'Z';
    return {
      id: r.id,
      name: r.name,
      mean: m,
      stddev: sd,
      cv,
      class: cls,
      totalSales: xs.reduce((s, x) => s + x, 0),
      periodsCount: xs.length,
    };
  });
}

/**
 * Совмещение ABC × XYZ — даёт матрицу 9 категорий.
 * AX — флагман (стабильно и прибыльно), CZ — кандидаты на вывод из ассортимента.
 */
export type AbcXyzCell = `${'A' | 'B' | 'C'}${'X' | 'Y' | 'Z'}`;

export function combineAbcXyz(
  abc: Array<{ id: string; class: 'A' | 'B' | 'C' }>,
  xyz: Array<{ id: string; class: 'X' | 'Y' | 'Z' }>,
): Map<string, AbcXyzCell> {
  const xyzMap = new Map(xyz.map((x) => [x.id, x.class]));
  const result = new Map<string, AbcXyzCell>();
  for (const a of abc) {
    const xClass = xyzMap.get(a.id);
    if (xClass) result.set(a.id, `${a.class}${xClass}` as AbcXyzCell);
  }
  return result;
}

File: backend/app/analytics/xyz.py
"""
XYZ анализ — порт widget/lib/analytics/xyz.ts.

Классификация товаров по стабильности спроса через коэффициент вариации
(CV = stddev / mean) на бакетах продаж.

  X: CV ≤ 0.10  (стабильный спрос, можно JIT)
  Y: 0.10 < CV ≤ 0.25  (умеренные колебания)
  Z: CV > 0.25  (нерегулярный, нужен страховой запас)

Краевые случаи (повторяют TS):
- Пустой массив periods   → mean=0, cv=Infinity, класс Z
- mean == 0 (все нули)    → cv=Infinity, класс Z
- Один ненулевой бакет    → высокий cv, класс Z
- Все элементы равны      → cv=0, класс X

Используется ВЫБОРОЧНОЕ стандартное отклонение (n-1) — так же как
lib/utils.ts:stddev. На бакетах из 8 значений это слегка завышает CV
по сравнению с популяционной формулой (n), но «корректнее» для прогноза.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal

from .utils import mean, stddev, safe_filter_finite


XyzClass = Literal["X", "Y", "Z"]


@dataclass(frozen=True)
class XyzInput:
    id: str
    name: str
    periods: list[float]


@dataclass
class XyzRow:
    id: str
    name: str
    mean: float
    stddev: float
    cv: float            # может быть math.inf
    class_: XyzClass
    total_sales: float
    periods_count: int


@dataclass
class XyzThresholds:
    x: float = 0.10
    y: float = 0.25


def build_xyz_report(
    inputs: list[XyzInput],
    thresholds: XyzThresholds | None = None,
) -> list[XyzRow]:
    th = thresholds or XyzThresholds()
    rows: list[XyzRow] = []
    for r in inputs:
        xs = safe_filter_finite(r.periods)
        m = mean(xs)
        sd = stddev(xs)
        cv = (sd / m) if m > 0 else math.inf
        cls: XyzClass = "X" if cv <= th.x else ("Y" if cv <= th.y else "Z")
        rows.append(
            XyzRow(
                id=r.id, name=r.name,
                mean=m, stddev=sd, cv=cv, class_=cls,
                total_sales=sum(xs),
                periods_count=len(xs),
            )
        )
    return rows
